### PR TITLE
Feat: Vehicle State Errors

### PR DESCRIPTION
**Pull request by adolfoportilla**

### CODE REVIEW

File: src/main/java/com/smartcar/sdk/AuthClient.java [hunk 1]
```diff
@@ -97,13 +97,12 @@ public static String getUserId(String accessToken) throws SmartcarException {
 
     try {
       json = new Gson().fromJson(response.body().string(), JsonObject.class);
+      return json.get("id").getAsString();
     } catch (IOException ex) {
       throw new SmartcarException(ex.getMessage());
-    } catch (NullPointerException ex) {
-      throw new SmartcarException("Received an empty response body.");
+    } catch (NullPointerException ex)  {
+      throw new SmartcarException(ex.getMessage());
```
Comment (gurpreetatwal) [inline src/main/java/com/smartcar/sdk/AuthClient.java]: should we just forward the null pointer exception here?

Comment (johnmryan) [reply src/main/java/com/smartcar/sdk/AuthClient.java]: Agreed with this, we probably shouldn't coerce an NPE into a `SmartcarException`

**Comment by johnmryan** [comment]
Just a quick thought based on the discussion we had earlier, can punt on it for now since it's probably out of scope.

```
public class ResponseHandler {
    public static handleErrorResponse(final Response response) throws IOException {
        JsonObject body = gson.fromJson(response.body().string(), JsonObject.class);
        String code = "";
        JsonElement codeElement = body.get("code");
        if (codeElement != null) {
            code = codeElement.getAsString();
        }
        String message = body.get("message").getAsString();
        String error = body.get("error").getAsString();
        Integer statusCode = response.code();
        return new ResponseError(statusCode, error, message, code);
    }
}

public class ResponseError {
    public ResponseError(int statusCode, String error, String message, int code) {
        ...
    }
}
```

```
if (!response.isSuccessful()) {
    ResponseError error = ResponseHandler.handleErrorResponse(response);
    throw new SmartcarException(error.getMessage(), error.getCode(), ...);
}
```

File: src/main/java/com/smartcar/sdk/SmartcarException.java [hunk 2]
```diff
@@ -98,107 +33,57 @@ public SmartcarException() {
...
+    JsonObject body = gson.fromJson(response.body().string(), JsonObject.class);
 
-  /**
-   * Returns the HTTP response associated with the exception.
-   *
-   * @return the response object
-   */
-  public Response getResponse() {
-    return this.response;
+    String code = "";
```
Comment (johnmryan) [inline src/main/java/com/smartcar/sdk/SmartcarException.java]: Do we want the default to be an empty string or `null`?

File: src/main/java/com/smartcar/sdk/SmartcarException.java [hunk 3]
```diff
@@ -98,107 +33,57 @@ public SmartcarException() {
...
+    JsonElement codeElement = body.get("code");
+    if (codeElement != null) {
+      code = codeElement.getAsString();
+    }
+    String message = body.get("message").getAsString();
+    String error = body.get("error").getAsString();
+    Integer statusCode = response.code();
```
Comment (johnmryan) [inline src/main/java/com/smartcar/sdk/SmartcarException.java]: Why `Integer` and not `int`? Is that what this returns?

File: src/main/java/com/smartcar/sdk/SmartcarException.java (hunk 2 above)
Comment (adolfoportilla) [reply src/main/java/com/smartcar/sdk/SmartcarException.java]: Agree, `null` its better

File: src/main/java/com/smartcar/sdk/SmartcarException.java (hunk 3 above)
Comment (adolfoportilla) [reply src/main/java/com/smartcar/sdk/SmartcarException.java]: should be `int`, done~

File: src/main/java/com/smartcar/sdk/AuthClient.java (hunk 1 above)
Comment (adolfoportilla) [reply src/main/java/com/smartcar/sdk/AuthClient.java]: Done, forgot to commit it, :upside_down_face:

**Pull request merged**